PRIME-1338 - Email Refactor

james-hollinger opened the pull request:
Made a 'canonical' model for Emails that are consumed by the CHES and SMTP clients. Lots of simplification.

neophyte57 commented:
@james-hollinger While I see that you've tagged the PR with `Needs Tests`, can you add some steps of how to test your refactoring or a link to such steps.  I also ran `dotnet test` in the `prime-dotnet-webapi-tests` project and get some compile errors; is that expected?

![image](https://user-images.githubusercontent.com/7586665/106048807-91adff80-60a2-11eb-837d-bcd47f134f22.png)

AnaisHebert reviewed (commented):
I got an error when updating remote users on a site when sending that email 
![image](https://user-images.githubusercontent.com/53796823/106073440-ecedeb00-60be-11eb-84d2-292c99bd7a56.png)

james-hollinger commented:
> 
> 
> @james-hollinger While I see that you've tagged the PR with `Needs Tests`, can you add some steps of how to test your refactoring or a link to such steps. I also ran `dotnet test` in the `prime-dotnet-webapi-tests` project and get some compile errors; is that expected?
> 
> ![image](https://user-images.githubusercontent.com/7586665/106048807-91adff80-60a2-11eb-837d-bcd47f134f22.png)

Yes, the backend tests are currently broken.

james-hollinger commented:
> 
> 
> I got an error when updating remote users on a site when sending that email
> ![image](https://user-images.githubusercontent.com/53796823/106073440-ecedeb00-60be-11eb-84d2-292c99bd7a56.png)

Thanks, I forgot to update the org agreement chtmls with the new view models

neophyte57 commented:
This `EmailsController` can be used for verifying that all types of emails are sent and formatted properly.  See the `TestSendAllEmails` method.

```
using System.Threading.Tasks;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;

using Prime.Auth;
using Prime.Models;
using Prime.Services;

using System.Linq;

namespace Prime.Controllers
{
    [Produces("application/json")]
    [Route("api/[controller]")]
    [ApiController]
    public class EmailsController : ControllerBase
    {
        private readonly IEmailService _emailService;
        private readonly ISiteService _siteService;
        private readonly IEnrolleeService _enrolleeService;

        public EmailsController(IEmailService emailService, ISiteService siteService, IEnrolleeService enrolleeService)
        {
            _emailService = emailService;
            _siteService = siteService;
            _enrolleeService = enrolleeService;
        }

        // POST: api/Emails/management/statuses
        /// <summary>
        /// Update all logged email statuses sent using the CHES email service
        /// </summary>
        [HttpPost("management/statuses", Name = nameof(UpdateEmailLogStatuses))]
        [Authorize(Roles = Roles.PrimeApiServiceAccount + "," + Roles.PrimeAdmin)]
        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
        [ProducesResponseType(StatusCodes.Status403Forbidden)]
        [ProducesResponseType(StatusCodes.Status204NoContent)]
        public async Task<StatusCodeResult> UpdateEmailLogStatuses()
        {
            await _emailService.UpdateEmailLogStatuses();

            return NoContent();
        }

        // POST: api/Emails/management/enrollees/renewal
        /// <summary>
        /// Send enrollee renewal reminder emails
        /// </summary>
        [HttpPost("management/enrollees/renewal", Name = nameof(SendEnrolleeRenewalEmails))]
        [Authorize(Roles = Roles.PrimeApiServiceAccount + "," + Roles.PrimeAdmin)]
        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
        [ProducesResponseType(StatusCodes.Status403Forbidden)]
        [ProducesResponseType(StatusCodes.Status204NoContent)]
        public async Task<ActionResult> SendEnrolleeRenewalEmails()
        {
            await _emailService.SendEnrolleeRenewalEmails();

            return NoContent();
        }

        // POST: api/Emails/management/test-send-all-emails
        /// <summary>
        /// Test sending of all possible emails.
        /// </summary>
        [HttpPost("management/test-send-all-emails", Name = nameof(TestSendAllEmails))]
        [ProducesResponseType(StatusCodes.Status204NoContent)]
        [ProducesResponseType(StatusCodes.Status400BadRequest)]
        public async Task<ActionResult> TestSendAllEmails()
        {
            var sites = await _siteService.GetSitesAsync();
            var site = sites.FirstOrDefault();
            if (site == null)
            {
                return StatusCode(StatusCodes.Status400BadRequest, "Please first submit a site using web UI (http://localhost:4200/site).");
            }
            var enrollees = await _enrolleeService.GetEnrolleesAsync();
            var enrollee = enrollees.FirstOrDefault();
            if (enrollee == null)
            {
                return StatusCode(StatusCodes.Status400BadRequest, "Please first submit an enrollee using web UI (http://localhost:4200/info).");
            }

            await _emailService.SendBusinessLicenceUploadedAsync(site);

            // Testing this requires some careful manipulation of the Agreement ExpiryDate for enrollees
            await _emailService.SendEnrolleeRenewalEmails();

            // DeviceProvider not currently handled
            for (int careSetting = 1; careSetting < 4; careSetting++)
            {
                var token = new EnrolmentCertificateAccessToken() { EnrolleeId = enrollee.Id };
                await _emailService.SendProvisionerLinkAsync(new string[] { "a@b.com" }, token, careSetting);
            }

            await _emailService.SendReminderEmailAsync(enrollee.Id);

            // Find site with some remote users
            site = sites.First(s => s.RemoteUsers.FirstOrDefault() != null);
            if (site == null)
            {
                return StatusCode(StatusCodes.Status400BadRequest, "Please first associate remote users with a site using web UI (http://localhost:4200/site).");
            }
            await _emailService.SendRemoteUserNotificationsAsync(site, site.RemoteUsers);

            await _emailService.SendRemoteUsersUpdatedAsync(site);

            await _emailService.SendSiteApprovedHIBCAsync(site);

            await _emailService.SendSiteApprovedPharmaNetAdministratorAsync(site);

            await _emailService.SendSiteApprovedSigningAuthorityAsync(site);

            await _emailService.SendSiteRegistrationSubmissionAsync(site.Id);

            return NoContent();
        }
    }
}
```

neophyte57 commented:
> > @james-hollinger While I see that you've tagged the PR with `Needs Tests`, can you add some steps of how to test your refactoring or a link to such steps. I also ran `dotnet test` in the `prime-dotnet-webapi-tests` project and get some compile errors; is that expected?
> > ![image](https://user-images.githubusercontent.com/7586665/106048807-91adff80-60a2-11eb-837d-bcd47f134f22.png)
> 
> Yes, the backend tests are currently broken.

I'm glad @james-hollinger that you got backend tests mostly working ... I added some code to `TestUtils.cs` to address the following: 

![image](https://user-images.githubusercontent.com/7586665/106308002-56840b80-621d-11eb-988e-d2949dd46f67.png)

james-hollinger commented:
> 
> 
> > > @james-hollinger While I see that you've tagged the PR with `Needs Tests`, can you add some steps of how to test your refactoring or a link to such steps. I also ran `dotnet test` in the `prime-dotnet-webapi-tests` project and get some compile errors; is that expected?
> > > ![image](https://user-images.githubusercontent.com/7586665/106048807-91adff80-60a2-11eb-837d-bcd47f134f22.png)
> > 
> > 
> > Yes, the backend tests are currently broken.
> 
> I'm glad @james-hollinger that you got backend tests mostly working ... I added some code to `TestUtils.cs` to address the following:
> 
> ![image](https://user-images.githubusercontent.com/7586665/106308002-56840b80-621d-11eb-988e-d2949dd46f67.png)

The error is actually in the application, that is a legitimate failing test. A null reference error when JSON serializing one of the properties, which I fixed in PR#1094.

As well, the TestUtils file is deprecated; I have been trying to use it less and less whenever I come back to fixing tests. The model factories we are using in our tests are in the ModelFactories folder.

neophyte57 commented:
A few things:

- I think you want the following for your test data to fully test the range of days before expiry:

![image](https://user-images.githubusercontent.com/7586665/106389599-51e86000-63a1-11eb-96ba-f71d43ab1a34.png)

- Can you add a `README.md` for prime-dotnet-webapi-tests that includes at least the following commands:  `dotnet test` and to show which tests ran and with what data `dotnet test -v n`

- Add a comment to `TestUtils` indicating something similar to "the TestUtils file is deprecated; I have been trying to use it less and less whenever I come back to fixing tests. The model factories we are using in our tests are in the ModelFactories folder."  😉

neophyte57 commented:
Although this is outside the scope of this refactoring, shouldn't there be more use of the `<p>` tag rather than `<br>` tag inside the  Razor templates?  E.g. `Views\Emails\HealthAuthorityEmail.cshtml` 

![image](https://user-images.githubusercontent.com/7586665/106403364-86810980-63eb-11eb-904b-7354276d4f12.png)

renders somewhat poorly visually because of the lack of paragraphs delimiters.

![image](https://user-images.githubusercontent.com/7586665/106403401-c2b46a00-63eb-11eb-8624-cdcff02d52b8.png)

mtpultz commented:
@neophyte57 yah definitely, we never ended up getting tickets for adding branding to the emails so definitely worth improving the semantics of the email markup, and maybe a base email razor partial with a small set of common styles.  Great low point technical backlog ticket if you could add it.

neophyte57 commented on prime-dotnet-webapi/Views/Emails/RemoteUsersUpdatedEmail.cshtml:
```diff
@@ -0,0 +1,28 @@
+
+<p>
+  Notification: The list of Remote Practitioners at @Model.SiteStreetAddress of
+  @Model.OrganizationName (PEC: @Model.SitePec) has been updated. The Remote Practitioners
```
Can `@Model.SitePec` be null?  

![image](https://user-images.githubusercontent.com/7586665/106486044-7a8a5b80-646e-11eb-92eb-4d9d3978b257.png)

neophyte57 commented on prime-dotnet-webapi-tests/Unit Tests/RazorConverterService.Tests.cs:
```diff
@@ -142,17 +234,16 @@ public async void TestRender_SiteSummary()
 
         [Theory]
         [MemberData(nameof(DocumentTemplates))]
-        public async void TestRender_Documents(RazorTemplate<Document> template)
+        public async void TestRender_Documents(RazorTemplate<File> template)
         {
             var service = CreateService();
-            var model = new Document("filename.ext", new byte[] { });
+            var model = new File("filename.ext", new byte[] { });
 
             var html = await service.RenderTemplateToStringAsync(template, model);
 
             Assert.NotNull(html);
         }
```
Would it be better to split apart this test, so that `DocumentTemplates` data is not used?  Then there can be a more substantial assertion than simply `Assert.NotNull(html);`?

neophyte57 commented on prime-dotnet-webapi/ViewModels/Agreements/OrgAgreementRazorViewModel.cs:
```diff
@@ -0,0 +1,16 @@
+
+namespace Prime.ViewModels.Agreements
+{
+    public class OrgAgreementRazorViewModel
+    {
```
Is there a need to include `Razor` in the classname?  Also seems inconsistent compared to other ViewModel classes

james-hollinger replied:
This is a ViewModel used by Razor specifically. It is used to render the Org Agreement.

neophyte57 commented on prime-dotnet-webapi-tests/Unit Tests/RazorConverterService.Tests.cs:
```diff
@@ -77,62 +81,150 @@ public async void TestRender_Agreement_WithLimits(RazorTemplate<Agreement> templ
         }
 
         [Fact]
         public async void TestRender_SiteSummary()
         {
```
Great that you've added all these unit tests but can `TestRender_SiteSummary` verify more of the generated `SiteRegistrationReview` content?

james-hollinger replied:
Most (all?) of the other razor templates have small ViewModels and use the data semantically; so we expect to see it in the rendered html with little variation.  The site registration review, however, is more of a collection of almost every piece of data about a site, and so is likely to change if we add or remove anything to the site Model. I was hesitant to test that any specific data was included in the rendered html lest the test becomes overly brittle.

neophyte57 replied:
It would just be nice to verify that some semblance of the expected document is produced

neophyte57 replied:
@james-hollinger As we talked about on Teams, I've updated `TestRender_SiteSummary`.  I also updated `prime-dotnet-webapi-tests\README.md` after testing CHES client

neophyte57 reviewed (approved):
LGTM  ✔
